### PR TITLE
fix(icon): Use data URLS instead of $templateCache

### LINKED ISSUE #6531: (Re)cache used icons like "md-tabs-arrow'" after $templateCache.removeAll()
Hi guys,

For my app we need to delete the entire templateCache after creating or updating some GUI forms. The problem is the MdIconProvider with function "preloadIcons" cached some icons in the templatecache.

After $templateCache.removeAll() the cached icons are gone ;)

Is there an option to rerun this "preloadIcons" function on runtime (in a controler)? 

Would be nice!

Grtz Dennis

### PR DESCRIPTION
**Pull request by programmist**
Fixes #6531

### CODE REVIEW

**Comment by programmist** [comment]
@ThomasBurleson: Could you have a look when you get a minute? Thanks!

File: src/components/icon/js/iconService.js [hunk 1]
```diff
@@ -1,5 +1,13 @@
   angular
     .module('material.components.icon' )
+    .constant('svgRegistry', {
```
Comment (DevVersion) [inline src/components/icon/js/iconService.js]: @programmist What do you think about privatizing that a bit more. Something like `$$mdSvgRegistry`?

For example, `$$interimElement` is also an internal service for ngMaterial, so I think the `$$` prefix should be used consistent. Thoughts?

Comment (programmist) [inline src/components/icon/js/iconService.js]: Good suggestion! I wouldn't want anyone accidentally trying to use this.

Comment (DevVersion) [inline src/components/icon/js/iconService.js]: Yes, that was my thought :)

Comment (programmist) [inline src/components/icon/js/iconService.js]: Done. Commit amended.

File: src/components/icon/js/iconService.js [hunk 2]
```diff
@@ -1,5 +1,13 @@
...
+        'mdCancel':      'data:image/svg+xml;base64,' + btoa('<svg version="1.1" x="0px" y="0px" viewBox="0 0 24 24"><g><path d="M12 2c-5.53 0-10 4.47-10 10s4.47 10 10 10 10-4.47 10-10-4.47-10-10-10zm5 13.59l-1.41 1.41-3.59-3.59-3.59 3.59-1.41-1.41 3.59-3.59-3.59-3.59 1.41-1.41 3.59 3.59 3.59-3.59 1.41 1.41-3.59 3.59 3.59 3.59z"/></g></svg>'),
+        'mdMenu':        'data:image/svg+xml;base64,' + btoa('<svg version="1.1" x="0px" y="0px" viewBox="0 0 24 24"><path d="M3,6H21V8H3V6M3,11H21V13H3V11M3,16H21V18H3V16Z" /></svg>'),
+        'mdToggleArrow': 'data:image/svg+xml;base64,' + btoa('<svg version="1.1" x="0px" y="0px" viewBox="0 0 48 48"><path d="M24 16l-12 12 2.83 2.83 9.17-9.17 9.17 9.17 2.83-2.83z"/><path d="M0 0h48v48h-48z" fill="none"/></svg>'),
+        'mdCalendar':    'data:image/svg+xml;base64,' + btoa('<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"><path d="M19 3h-1V1h-2v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 16H5V8h14v11zM7 10h5v5H7z"/></svg>')
```
Comment (Frank3K) [inline src/components/icon/js/iconService.js]: While this is good for readability for developers, would you want to let the browser do this?

**Comment by Frank3K** [comment]
Just wondering, is there any performance impact when changing from template cache to base64 data? In my experience base64 is slower, probably because the browser needs to perform a decode step and maybe because it is less optimized since it's not as widely used.

File: src/components/icon/js/iconService.js (hunk 2 above)
Comment (programmist) [inline src/components/icon/js/iconService.js]: I'm not too concerned with it, but I do agree that using escaped, unencoded strings would be better. I'll submit that change shortly.

**Comment by ThomasBurleson** [comment]
@programmist - is this PR still valid and ready ?

**Comment by programmist** [comment]
@ThomasBurleson: It was probably ok as-is, but I'll went ahead and pre-encoded those SVG strings rather than relying on the browser to do it at runtime.

**Comment by ThomasBurleson** [comment]
@programmist - Thx Tony.

**Comment by ThomasBurleson** [comment]
@programmist  - can you rebase with `master` please ?

**Comment by programmist** [comment]
@ThomasBurleson: Sure, I'll comment here once rebase is done.

**Comment by programmist** [comment]
@ThomasBurleson: Rebased and ready to merge at your convenience.

**Pull request closed without merging**